src/algebra/lie/basic.lean

```diff
@@ -42,119 +41,96 @@ lie bracket, ring commutator, jacobi identity, lie ring, lie algebra
 
 universes u v w w₁
 
-/-- A binary operation, intended use in Lie algebras and similar structures. -/
-class has_bracket (L : Type v) := (bracket : L → L → L)
+/-- The has_bracket class has two intended uses:
 
-notation `⁅`x`,` y`⁆` := has_bracket.bracket x y
-
-/-- An Abelian Lie algebra is one in which all brackets vanish. Arguably this class belongs in the
-`has_bracket` namespace but it seems much more user-friendly to compromise slightly and put it in
-the `lie_algebra` namespace. -/
-class lie_algebra.is_abelian (L : Type v) [has_bracket L] [has_zero L] : Prop :=
-(abelian : ∀ (x y : L), ⁅x, y⁆ = 0)
-
-namespace ring_commutator
-
-variables {A : Type v} [ring A]
-
-/-- The bracket operation for rings is the ring commutator, which captures the extent to which a
-ring is commutative. It is identically zero exactly when the ring is commutative. -/
-@[priority 100]
-instance : has_bracket A :=
-{ bracket := λ x y, x*y - y*x }
+  1. for the product `⁅x, y⁆` of two elements `x`, `y` in a Lie algebra
+     (analogous to `has_mul` in the associative setting),
 
-lemma commutator (x y : A) : ⁅x, y⁆ = x*y - y*x := rfl
+  2. for the action `⁅x, m⁆` of an element `x` of a Lie algebra on an element `m` in one of its
+     modules (analogous to `has_scalar` in the associative setting). -/
+class has_bracket (L : Type v) (M : Type w) := (bracket : L → M → M)
 
-end ring_commutator
+notation `⁅`x`,` y`⁆` := has_bracket.bracket x y
 
 /-- A Lie ring is an additive group with compatible product, known as the bracket, satisfying the
 Jacobi identity. The bracket is not associative unless it is identically zero. -/
-@[protect_proj] class lie_ring (L : Type v) extends add_comm_group L, has_bracket L :=
+@[protect_proj] class lie_ring (L : Type v) extends add_comm_group L, has_bracket L L :=
 (add_lie : ∀ (x y z : L), ⁅x + y, z⁆ = ⁅x, z⁆ + ⁅y, z⁆)
-(lie_add : ∀ (x y z : L), ⁅z, x + y⁆ = ⁅z, x⁆ + ⁅z, y⁆)
+(lie_add : ∀ (x y z : L), ⁅x, y + z⁆ = ⁅x, y⁆ + ⁅x, z⁆)
 (lie_self : ∀ (x : L), ⁅x, x⁆ = 0)
-(jacobi : ∀ (x y z : L), ⁅x, ⁅y, z⁆⁆ + ⁅y, ⁅z, x⁆⁆ + ⁅z, ⁅x, y⁆⁆ = 0)
+(lie_lie : ∀ (x y z : L), ⁅⁅x, y⁆, z⁆ = ⁅x, ⁅y, z⁆⁆ - ⁅y, ⁅x, z⁆⁆)
```

Comment:
Looking now I wonder if the following alternative to `lie_lie` would be preferable:
```lean
(leibinz_lie : ∀ (x y z : L),  ⁅x, ⁅y, z⁆⁆ = ⁅⁅x, y⁆, z⁆ + ⁅y, ⁅x, z⁆⁆)
```

Reply:
In the end, it doesn't really matter, right? We will need all variations.
We might want to provide several different constructors...

Reply:
Yeah, it doesn't really of course.

I was just thinking it's slightly nicer not to require subtraction when its not necessary. And who knows, maybe some day we'll enter the exotic / somewhat hilarious world of Lie rings over monoids that are not necessarily groups?

Reply:
Ok, let's go with `leibniz_lie`.

Reply:
Done!